core/gesture.js

```diff
@@ -0,0 +1,545 @@
+    // Click the workspace.
+    this.doWorkspaceClick_();
+  }
+  this.endGesture(e);
+};
+
+/**
+ * End the gesture associated with the given event, and remove it from the
+ * gesture database.
+ * @param {!Event} e A mouse move or touch move event.
+ */
+Blockly.Gesture.prototype.endGesture = function(e) {
+  Blockly.GestureDB.removeGestureForId(this.touchIdentifier_);
+  e.preventDefault();
+  e.stopPropagation();
+};
+
+/**
+ * Handle a right-click event by showing a context menu.
+ * @param {!Event} e A mouse move or touch move event.
+ */
+Blockly.Gesture.prototype.handleRightClick = function(e) {
+  if (this.startBlock_) {
+    this.startBlock_.showContextMenu_(e);
+  } else if (this.startWorkspace_) {
+    this.startWorkspace_.showContextMenu_(e);
+  }
+  this.endGesture(e);
+};
+
+// Called externally.
+/**
+ * Handle a mousedown/touchstart event on a workspace.
+ * @param {!Event} e A mouse down or touch start event.
+ * @param {!Blockly.Workspace} ws The workspace the event hit.
+ */
+Blockly.Gesture.prototype.handleWsStart = function(e, ws) {
+  this.setStartWorkspace(ws);
+  this.mostRecentEvent_ = e;
+  this.doStart(e);
+};
+
+/**
+ * Handle a mousedown/touchstart event on a flyout.
+ * @param {!Event} e A mouse down or touch start event.
+ * @param {!Blockly.Flyout} flyout The flyout the event hit.
+ */
+Blockly.Gesture.prototype.handleFlyoutStart = function(e, flyout) {
+  this.setStartFlyout(flyout);
+  this.handleWsStart(e, flyout.getWorkspace());
+};
+
+/**
+ * Handle a mousedown/touchstart event on a block.
+ * @param {!Event} e A mouse down or touch start event.
+ * @param {!Blockly.BlockSvg} block The block the event hit.
+ */
+Blockly.Gesture.prototype.handleBlockStart = function(e, block) {
+  this.setStartBlock(block);
+  this.mostRecentEvent_ = e;
+};
+
+// Field clicks
+Blockly.Gesture.prototype.doFieldClick_ = function() {
+  this.startField_.showEditor_();
+};
+
+// Block clicks
+Blockly.Gesture.prototype.doBlockClick_ = function() {
+  // TODO: Implement.
+};
+
+// Workspace clicks
+Blockly.Gesture.prototype.doWorkspaceClick_ = function() {
+  // TODO: Implement.
+};
+
+// Simple setters
+
+/**
+ * Record the field that a gesture started on.
+ * @param {Blockly.Field} field The field the gesture started on.
+ */
+Blockly.Gesture.prototype.setStartField = function(field) {
+  if (!this.startField_) {
+    this.startField_ = field;
+  }
+};
+
+/**
+ * Record the block that a gesture started on.
+ * @param {Blockly.BlockSvg} block The block the gesture started on.
+ */
+Blockly.Gesture.prototype.setStartBlock = function(block) {
+  if (!this.startBlock_) {
+    if (block.isShadow()) {
+      this.setStartBlock(block.parentBlock_);
+    } else {
+      this.startBlock_ = block;
+    }
+  }
+};
+
+/**
+ * Record the workspace that a gesture started on.
+ * @param {Blockly.WorkspaceSvg} ws The workspace the gesture started on.
+ */
+Blockly.Gesture.prototype.setStartWorkspace = function(ws) {
```

Comment:
I think that could only happen if we have an error in our event handling.  (So yes, it's expected.)

The case I had in mind was: mouse down in always-open flyout, a failure to call stopPropagation, and then mouse down firing in the main workspace as well.